File: python/pet/ui/fun_image_popup.py
# -*- coding: utf-8 -*-
"""Frameless, stackable image windows for the playful menu entry."""
from __future__ import annotations

from pet.infrastructure.paths import resource_root

import os
import random
from pathlib import Path
from typing import Callable

from PySide6.QtCore import QPoint, Qt
from PySide6.QtGui import QCloseEvent, QImageReader, QMouseEvent, QPixmap
from PySide6.QtWidgets import (
    QApplication,
    QHBoxLayout,
    QLabel,
    QPushButton,
    QVBoxLayout,
    QWidget,
)


def oijingjing_image_path() -> Path:
    return resource_root() / "assets" / "big_blue_fat_fish" / "ojingjing.jpg"


def bundled_assets_root() -> Path:
    """应用内置资产根目录（源码树或 PyInstaller _internal）。"""
    return resource_root() / "assets"


def resolve_fun_asset(path: str | Path | None, fallback: Path) -> Path:
    fallback = Path(fallback).expanduser()
    if path is None or not str(path).strip():
        return fallback
    candidate = Path(str(path or "")).expanduser()
    if candidate.is_absolute():
        return candidate if candidate.exists() else fallback
    bundled = resource_root() / candidate
    return bundled if bundled.exists() else fallback


def store_fun_asset(value, default: Path | str) -> str:
    """把路径转成适合持久化的形式（保持 portable）。

    - 应用内置 assets 内的路径（含用户曾保存的绝对路径）→ 存相对值
      assets/...，目录移动/自更新后仍可解析；
    - 用户自选的外部文件 → 存绝对路径原样保留。
    """
    default = Path(default)
    candidate = str(value or "").strip()
    if not candidate:
        return str(default)
    path = Path(candidate).expanduser()
    if not path.is_absolute():
        # Windows 上统一正斜杠，避免配置里混入反斜杠（跨平台一致）
        return candidate.replace("\\", "/") if os.name == "nt" else candidate
    try:
        rel = path.resolve().relative_to(bundled_assets_root().resolve())
        # 统一正斜杠：配置值与 legacy 迁移比较、跨平台一致
        return str(Path("assets") / rel).replace("\\", "/")
    except ValueError:
        return str(path)


def _readable_image_paths(directory: Path) -> list[Path]:
    if not directory.is_dir():
        return []
    supported = {".jpg", ".jpeg", ".png", ".webp", ".bmp", ".gif", ".tif", ".tiff"}
    try:
        return sorted(
            path for path in directory.iterdir()
            if path.is_file()
            and path.suffix.lower() in supported
            and QImageReader(str(path)).canRead()
        )
    except OSError:
        return []


def popup_image_paths(directory: str | Path | None = None) -> list[Path]:
    fallback = oijingjing_image_path().parent
    requested = resolve_fun_asset(directory, fallback)
    paths = _readable_image_paths(requested)
    if paths or requested == fallback:
        return paths
    return _readable_image_paths(fallback)


class OjingjingWindow(QWidget):
    def __init__(
        self,
        image_path: Path,
        close_all: Callable[[], None],
        on_closed: Callable[[QWidget], None],
        title: str = "厉害了我的鲸",
    ) -> None:
        super().__init__(None)
        self._on_closed = on_closed
        self._drag_start: QPoint | None = None
        self._window_start: QPoint | None = None
        self.setObjectName("ojingjingImageWindow")
        self.setProperty("sourceImage", str(image_path))
        self.setWindowTitle(str(title or "厉害了我的鲸"))
        self.setWindowFlags(
            Qt.WindowType.Window
            | Qt.WindowType.FramelessWindowHint
            | Qt.WindowType.WindowStaysOnTopHint
        )
        self.setAttribute(Qt.WidgetAttribute.WA_DeleteOnClose, True)
        self.setAttribute(Qt.WidgetAttribute.WA_TranslucentBackground, True)

        screen = QApplication.primaryScreen()
        available = screen.availableGeometry() if screen is not None else None
        side = 520 if available is None else max(280, min(560, int(available.height() * 0.64)))

        card = QWidget(self)
        card.setObjectName("ojingjingCard")
        card.setStyleSheet(
            "#ojingjingCard { background: white; border: 1px solid #d8d8d8; "
            "border-radius: 16px; }"
            "QPushButton { min-height: 30px; padding: 0 16px; border: 1px solid #d5d5d5; "
            "border-radius: 8px; background: #f7f7f7; color: #222; font-size: 13px; }"
            "QPushButton:hover { background: #ececec; }"
            "QPushButton#closeAllButton { background: #202020; color: white; border-color: #202020; }"
        )
        card_layout = QVBoxLayout(card)
        card_layout.setContentsMargins(10, 10, 10, 10)
        card_layout.setSpacing(10)

        image = QLabel(card)
        image.setObjectName("ojingjingFullImage")
        image.setAlignment(Qt.AlignmentFlag.AlignCenter)
        source = QPixmap(str(image_path))
        if source.isNull():
            raise ValueError(f"无法读取彩蛋图片: {image_path}")
        image.setPixmap(
            source.scaled(
                side,
                side,
                Qt.AspectRatioMode.KeepAspectRatio,
                Qt.TransformationMode.SmoothTransformation,
            )
        )
        image.setFixedSize(side, side)
        image.setAttribute(Qt.WidgetAttribute.WA_TransparentForMouseEvents, True)
        card_layout.addWidget(image)

        controls = QHBoxLayout()
        controls.addStretch(1)
        close_button = QPushButton("关闭", card)
        close_button.setObjectName("closeButton")
        close_button.clicked.connect(self.close)
        controls.addWidget(close_button)
        close_all_button = QPushButton("全部关闭", card)
        close_all_button.setObjectName("closeAllButton")
        close_all_button.clicked.connect(close_all)
        controls.addWidget(close_all_button)
        card_layout.addLayout(controls)

        root = QVBoxLayout(self)
        root.setContentsMargins(8, 8, 8, 8)
        root.addWidget(card)
        # Avoid QWidget.adjustSize() here: on the Cocoa plugin a freshly
        # created translucent frameless tool window can crash while querying
        # the native frame. Its content dimensions are deterministic.
        self.setFixedSize(side + 36, side + 82)

    def begin_drag_at(self, global_pos: QPoint) -> None:
        self._drag_start = QPoint(global_pos)
        self._window_start = self.pos()

    def drag_to(self, global_pos: QPoint) -> None:
        if self._drag_start is None or self._window_start is None:
            return
        self.move(self._window_start + global_pos - self._drag_start)

    def end_drag(self) -> None:
        self._drag_start = None
        self._window_start = None

    def mousePressEvent(self, event: QMouseEvent) -> None:  # noqa: N802 - Qt API
        if event.button() == Qt.MouseButton.LeftButton:
            self.begin_drag_at(event.globalPosition().toPoint())
            event.accept()
            return
        super().mousePressEvent(event)

    def mouseMoveEvent(self, event: QMouseEvent) -> None:  # noqa: N802 - Qt API
        if self._drag_start is not None and event.buttons() & Qt.MouseButton.LeftButton:
            self.drag_to(event.globalPosition().toPoint())
            event.accept()
            return
        super().mouseMoveEvent(event)

    def mouseReleaseEvent(self, event: QMouseEvent) -> None:  # noqa: N802 - Qt API
        if event.button() == Qt.MouseButton.LeftButton and self._drag_start is not None:
            self.drag_to(event.globalPosition().toPoint())
            self.end_drag()
            event.accept()
            return
        super().mouseReleaseEvent(event)

    def closeEvent(self, event: QCloseEvent) -> None:  # noqa: N802 - Qt API
        self._on_closed(self)
        super().closeEvent(event)


class OjingjingWindowManager:
    def __init__(self) -> None:
        self.windows: list[OjingjingWindow] = []

    def _forget(self, window: QWidget) -> None:
        self.windows = [item for item in self.windows if item is not window]

    def open_window(self, *, show: bool = True, config: dict | None = None) -> OjingjingWindow:
        config = dict(config or {})
        paths = popup_image_paths(config.get("image_dir"))
        if not paths:
            raise FileNotFoundError(f"没有可读取的彩蛋图片: {oijingjing_image_path().parent}")
        if show:
            self.restore_all()
        window = OjingjingWindow(
            random.choice(paths), self.close_all, self._forget,
            title=str(config.get("title") or "厉害了我的鲸"),
        )
        offset = (len(self.windows) % 7) * 24
        window.setProperty("cascadeOffset", offset)
        screen = QApplication.primaryScreen()
        if screen is not None:
            available = screen.availableGeometry()
            x = available.center().x() - window.width() // 2 + offset
            y = available.center().y() - window.height() // 2 + offset
            window.move(x, y)
        self.windows.append(window)
        if show:
            window.show()
            window.raise_()
            window.activateWindow()
        return window

    def restore_all(self) -> None:
        for window in self.windows:
            window.show()
            window.raise_()
        if self.windows:
            self.windows[-1].activateWindow()

    def close_all(self) -> None:
        for window in list(self.windows):
            window.close()
        self.windows.clear()


_MANAGER = OjingjingWindowManager()


def open_ojingjing_window(config: dict | None = None) -> OjingjingWindow:
    return _MANAGER.open_window(config=config)


def restore_ojingjing_windows() -> None:
    _MANAGER.restore_all()

File: c++/src/ui/FunImagePopup.cpp
#include "FunImagePopup.hpp"

#include "infrastructure/PathHelper.hpp"

#include <QApplication>
#include <QCloseEvent>
#include <QDir>
#include <QFileInfo>
#include <QHBoxLayout>
#include <QImageReader>
#include <QLabel>
#include <QMouseEvent>
#include <QPixmap>
#include <QPushButton>
#include <QRandomGenerator>
#include <QScreen>
#include <QVBoxLayout>
#include <algorithm>

namespace Pet::UI {

namespace {

const char* kWindowStyleSheet =
    "#ojingjingCard { background: white; border: 1px solid #d8d8d8; border-radius: 16px; }"
    "QPushButton { min-height: 30px; padding: 0 16px; border: 1px solid #d5d5d5; "
    "border-radius: 8px; background: #f7f7f7; color: #222; font-size: 13px; }"
    "QPushButton:hover { background: #ececec; }"
    "QPushButton#closeAllButton { background: #202020; color: white; border-color: #202020; }";

QStringList readableImagePaths(const QString& directory) {
    QDir dir(directory);
    if (!dir.exists()) {
        return {};
    }
    static const QStringList kSupported = {
        QStringLiteral("*.jpg"), QStringLiteral("*.jpeg"), QStringLiteral("*.png"),
        QStringLiteral("*.webp"), QStringLiteral("*.bmp"), QStringLiteral("*.gif"),
        QStringLiteral("*.tif"), QStringLiteral("*.tiff"),
    };
    QStringList result;
    const QFileInfoList entries = dir.entryInfoList(kSupported, QDir::Files, QDir::Name);
    for (const QFileInfo& info : entries) {
        if (QImageReader(info.absoluteFilePath()).canRead()) {
            result.append(info.absoluteFilePath());
        }
    }
    return result;
}

} // namespace

// ---------------------------------------------------------------- FunAssets

QString FunAssets::oijingjingImagePath() {
    return QDir(Infrastructure::PathHelper::getAssetsDir())
        .filePath(QStringLiteral("big_blue_fat_fish/ojingjing.jpg"));
}

QStringList FunAssets::popupImagePaths(const QString& configuredDir) {
    const QString fallbackDir = QFileInfo(oijingjingImagePath()).absolutePath();
    const QString requested = Infrastructure::PathHelper::resolveConfiguredAsset(
        configuredDir, fallbackDir);
    const QStringList paths = readableImagePaths(requested);
    if (!paths.isEmpty() || QDir::cleanPath(requested) == QDir::cleanPath(fallbackDir)) {
        return paths;
    }
    return readableImagePaths(fallbackDir);
}

// ---------------------------------------------------------------- OjingjingWindow

OjingjingWindow::OjingjingWindow(const QString& imagePath, const QString& title, QWidget* parent)
    : QWidget(parent) {
    setObjectName(QStringLiteral("ojingjingImageWindow"));
    setProperty("sourceImage", imagePath);
    setWindowTitle(title.isEmpty() ? QString::fromUtf8(u8"厉害了我的鲸") : title);
    setWindowFlags(Qt::Window | Qt::FramelessWindowHint | Qt::WindowStaysOnTopHint);
    setAttribute(Qt::WA_DeleteOnClose, true);
    setAttribute(Qt::WA_TranslucentBackground, true);

    QScreen* screen = QApplication::primaryScreen();
    const QRect available = screen != nullptr ? screen->availableGeometry() : QRect();
    const int side = available.isNull()
        ? 520
        : std::max(280, std::min(560, static_cast<int>(available.height() * 0.64)));

    auto* card = new QWidget(this);
    card->setObjectName(QStringLiteral("ojingjingCard"));
    card->setStyleSheet(QString::fromLatin1(kWindowStyleSheet));

    auto* cardLayout = new QVBoxLayout(card);
    cardLayout->setContentsMargins(10, 10, 10, 10);
    cardLayout->setSpacing(10);

    auto* image = new QLabel(card);
    image->setObjectName(QStringLiteral("ojingjingFullImage"));
    image->setAlignment(Qt::AlignCenter);
    const QPixmap source(imagePath);
    if (!source.isNull()) {
        image->setPixmap(source.scaled(side, side, Qt::KeepAspectRatio, Qt::SmoothTransformation));
    }
    image->setFixedSize(side, side);
    image->setAttribute(Qt::WA_TransparentForMouseEvents, true);
    cardLayout->addWidget(image);

    auto* controls = new QHBoxLayout();
    controls->addStretch(1);
    auto* closeButton = new QPushButton(QString::fromUtf8(u8"关闭"), card);
    closeButton->setObjectName(QStringLiteral("closeButton"));
    QObject::connect(closeButton, &QPushButton::clicked, this, &QWidget::close);
    controls->addWidget(closeButton);

    auto* closeAllButton = new QPushButton(QString::fromUtf8(u8"全部关闭"), card);
    closeAllButton->setObjectName(QStringLiteral("closeAllButton"));
    QObject::connect(closeAllButton, &QPushButton::clicked, this, []() {
        OjingjingWindowManager::instance().closeAll();
    });
    controls->addWidget(closeAllButton);
    cardLayout->addLayout(controls);

    auto* root = new QVBoxLayout(this);
    root->setContentsMargins(8, 8, 8, 8);
    root->addWidget(card);
    // 内容尺寸是确定的，避免在无边框半透明窗口上调用 adjustSize()。
    setFixedSize(side + 36, side + 82);
}

void OjingjingWindow::mousePressEvent(QMouseEvent* event) {
    if (event->button() == Qt::LeftButton) {
        m_dragging = true;
        m_dragStart = event->globalPosition().toPoint();
        m_windowStart = pos();
        event->accept();
        return;
    }
    QWidget::mousePressEvent(event);
}

void OjingjingWindow::mouseMoveEvent(QMouseEvent* event) {
    if (m_dragging && (event->buttons() & Qt::LeftButton)) {
        move(m_windowStart + event->globalPosition().toPoint() - m_dragStart);
        event->accept();
        return;
    }
    QWidget::mouseMoveEvent(event);
}

void OjingjingWindow::mouseReleaseEvent(QMouseEvent* event) {
    if (event->button() == Qt::LeftButton && m_dragging) {
        move(m_windowStart + event->globalPosition().toPoint() - m_dragStart);
        m_dragging = false;
        event->accept();
        return;
    }
    QWidget::mouseReleaseEvent(event);
}

void OjingjingWindow::closeEvent(QCloseEvent* event) {
    OjingjingWindowManager::instance().forget(this);
    QWidget::closeEvent(event);
}

// ---------------------------------------------------------------- OjingjingWindowManager

OjingjingWindowManager& OjingjingWindowManager::instance() {
    static OjingjingWindowManager manager;
    return manager;
}

void OjingjingWindowManager::forget(QWidget* window) {
    m_windows.erase(std::remove(m_windows.begin(), m_windows.end(), window), m_windows.end());
}

void OjingjingWindowManager::openWindow(const QString& configuredImageDir, const QString& title) {
    const QStringList paths = FunAssets::popupImagePaths(configuredImageDir);
    if (paths.isEmpty()) {
        qWarning("彩蛋图片目录中没有可读取的图片: %s",
                 qUtf8Printable(FunAssets::oijingjingImagePath()));
        return;
    }

    restoreAll();
    const int index = QRandomGenerator::global()->bounded(paths.size());
    auto* window = new OjingjingWindow(
        paths.at(index),
        title.isEmpty() ? QString::fromUtf8(u8"厉害了我的鲸") : title);

    const int offset = static_cast<int>(m_windows.size() % 7) * 24;
    window->setProperty("cascadeOffset", offset);
    if (QScreen* screen = QApplication::primaryScreen()) {
        const QRect available = screen->availableGeometry();
        window->move(available.center().x() - window->width() / 2 + offset,
                     available.center().y() - window->height() / 2 + offset);
    }
    m_windows.push_back(window);
    window->show();
    window->raise();
    window->activateWindow();
}

void OjingjingWindowManager::restoreAll() {
    for (QWidget* window : m_windows) {
        if (window == nullptr) {
            continue;
        }
        window->show();
        window->raise();
    }
    if (!m_windows.empty() && m_windows.back() != nullptr) {
        m_windows.back()->activateWindow();
    }
}

void OjingjingWindowManager::closeAll() {
    const std::vector<QWidget*> windows = m_windows;
    m_windows.clear();
    for (QWidget* window : windows) {
        if (window != nullptr) {
            window->close();
        }
    }
}

void openOjingjingWindow(const QString& configuredImageDir, const QString& title) {
    OjingjingWindowManager::instance().openWindow(configuredImageDir, title);
}

} // namespace Pet::UI
